exchanges/request/limit_test.go

```diff
@@ -0,0 +1,262 @@
+package request
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"testing/synctest"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"golang.org/x/time/rate"
+)
+
+func TestRateLimit(t *testing.T) {
+	t.Parallel()
+
+	synctest.Test(t, func(t *testing.T) { //nolint:thelper // false positive
+		err := (*RateLimiterWithWeight)(nil).RateLimit(t.Context())
+		assert.ErrorContains(t, err, "nil pointer: *request.RateLimiterWithWeight")
+
+		r := &RateLimiterWithWeight{limiter: rate.NewLimiter(rate.Limit(1), 1)}
+		err = r.RateLimit(t.Context())
+		assert.ErrorIs(t, err, errInvalidWeight, "should return errInvalidWeightCount for zero weight")
+
+		r = NewRateLimitWithWeight(time.Second, 10, 1)
+		start := time.Now()
+		err = r.RateLimit(t.Context())
+		elapsed := time.Since(start)
+		require.NoError(t, err, "rate limit must not error")
+		assert.Less(t, elapsed, time.Millisecond*50, "should complete quickly for first request")
+
+		r = NewRateLimitWithWeight(time.Second, 10, 5)
+		start = time.Now()
+		err = r.RateLimit(t.Context())
+		elapsed = time.Since(start)
+		require.NoError(t, err, "rate limit must not error")
+		assert.Less(t, elapsed, time.Millisecond*600, "should complete within reasonable time for weight 5")
+
+		r = NewRateLimitWithWeight(100*time.Millisecond, 1, 1)
+		start = time.Now()
+		err = r.RateLimit(WithDelayNotAllowed(t.Context()))
+		require.NoError(t, err, "first rate limit call must not error and must be immediate")
+		elapsed = time.Since(start)
+		assert.Less(t, elapsed, 50*time.Millisecond, "first call should be immediate")
+
+		start = time.Now()
+		err = r.RateLimit(t.Context())
+		require.NoError(t, err, "second rate limit call must not error")
+		elapsed = time.Since(start)
+		assert.GreaterOrEqual(t, elapsed, 90*time.Millisecond, "second call should be delayed by approximately 100ms")
+		assert.Less(t, elapsed, 150*time.Millisecond, "delay should not be excessive")
+
+		err = r.RateLimit(WithDelayNotAllowed(t.Context()))
+		assert.ErrorIs(t, err, ErrDelayNotAllowed, "should return correct error")
+
+		var routineErr error
+		wg := sync.WaitGroup{}
+		wg.Go(func() { routineErr = r.RateLimit(t.Context()) })
+		ctx, cancel := context.WithCancel(t.Context())
+		cancel()
+		time.Sleep(10 * time.Millisecond)
+		err = r.RateLimit(ctx)
+		assert.ErrorIs(t, err, context.Canceled, "should return correct error")
+		wg.Wait()
+		assert.NoError(t, routineErr, "waitgroup should be Done when context is canceled")
```

Comment:
🚧 Change request:
At the least, the comment here is incorrect.
But the more I look at Waitgroup usage here, the more I'm sure we're not using `synctest.Wait` when we should.

Reply:
This is what I came up with: gbjk@e8eae9a2c
```
curl -s https://github.com/gbjk/gocryptotrader/commit/e8eae9a2c.patch | git apply
```

Note: I might be missing something, but I couldn't see any value of the background RateLimits.
They're not testing anything we didn't test before. The contexts aren't inter-mingled.
I couldn't grok what they're adding _now_ other than complexity.
(Before maybe they added something, though)